### ISSUE TITLE
Create branch for Kubernetes CI testing

### ISSUE THREAD

**Issue by leblancd** [Mirantis/kubeadm-dind-cluster#115]
We would like to have a branch created off of Mirantis/k-d-c master that we will be used as a stable branch for using with Kubernetes CI testing. The goals of having a separate, stable branch are:
- The CI stable branch won't be (instantly) subject to changes introduced into master. We would rebase to master (and test) periodically/intermittently/as-needed.
- The CI stable branch would introduce a pointer to a pre-built node DinD image that is built from that branch. This pointer might be in the form of a text file that contains a DockerHub URL, or a shell script that sets the DinD image environment variable (I'm open to suggestions on this). The goal would be for the CI to be able to reference a fixed k-d-c branch on GitHub (we would never have to change the kubernetes/test-infra CI code to affect a k-d-c upgrade), and from this branch, the CI would clone/find matching DinD scripts and pointer to a pre-built DockerHub node DinD image.

I'm open to suggestions on the best way to implement this!
Thanks!

**Comment by ivan4th** [comment]
We can do `stable` branch for this, as well as tagging. In any case, current image build mechanism needs to be redone as it depends on Docker Cloud. This part of Docker Cloud is not going to be shut down right now IIRC, but overall I'm not very happy with the need to deal with its ui every time I update kubernetes versions etc. Travis CI has one serious drawback, if you use it in debug mode it will give up all the secrets that are used to push images etc. So we need to either (a) move to CircleCI (b) have both Travis and CircleCI build jobs, with CircleCI pushing the images. I think the latter is probably better.

**Comment by leblancd** [comment]
@ivan4th - Thanks for looking at this! Our group isn't using Travis CI nor CircleCI. Instead, we're using the standard Kubernetes test-infra Prow cluster, and we're having the Prow container/pod create a remote cluster on GCE using the gce-setup.sh script (via docker-machine). In our particular case, we're trying to test IPv6 on Kubernetes, but the test can be adapted to any test that wants to use a containerized cluster on GCE.

Here is [the PR for the IPv6 CI](https://github.com/kubernetes/test-infra/pull/7529)

In the current proposed CI, we pull the k-d-c "k8s-1.10" branch, using that as our stable branch:
https://github.com/kubernetes/test-infra/pull/7529/files#diff-406185368ba7839d1459d3d51424f104R1961

and we point to a manually-built node DinD image (built and pushed by @rpothier ):
https://github.com/kubernetes/test-infra/pull/7529/files#diff-5539d275f091e62154f51653bbe4fba0R12461

And that setting is used as a "DIND_IMAGE" environmental variable for gce_setup.sh:
https://github.com/kubernetes/test-infra/pull/7529/files#diff-19bf1416db48c9c35ab71f2cd9cd79cfR379

What we'd like to do is to consolidate this logic, so that the pulls a k-d-c stable branch (instead of k8s-1.10), and from that branch, it can somehow find the value that it needs for "DIND_IMAGE" environmental variable to find a DinD image that was built from the k-d-c stable branch. I suppose it can find a URL for the "DIND_IMAGE" in a text file, or a new "fixed" script can be created for this k-d-c stable branch, and that text file or script gets bumped whenever the stable branch is updated.

In other words, we were not expecting that the CI would do that image building, but rather we'd manually build when the stable branch is updated.

**Comment by ivan4th** [comment]
Sorry for delay, there was some breakage after the changes related to IPv6. k8s-1.10 branch is obsolete right now... Does current master effba5c772c08db01d5722618d7b323a0f45702d suit you for the stable branch? There's a new problem found, #110, that will be fixed soon, but it shouldn't affect you if you're not  using swap.

**Comment by leblancd** [comment]
Yes, current master should work for us, thanks! Re. swap, we usually manually disable that on our local hosts, and I'm pretty sure that GCE instances have swaps disabled.
Is there an easy way to solve the 2nd bullet in the description above? I don't know if I explained it well, but we had been using the k8s-1.10 branch and the corresponding k8s-1.10 DinD node image for CI. For the k8s-1.10 DinD node image, we used a fixed URL to the DockerHub image (i.e. DockerHub URL was compiled into kubernetes/test-infra). For this new stable branch, it would be better if the stable branch itself had a pointer (or script) to the latest DinD node image build from that branch. That way, our kubernetes/test-infra CI would never need to be modified... it would be able to find the DinD node image from the stable branch.

**Comment by ivan4th** [comment]
Yes, I think we can add `stable` script in fixed/ directory with `EMBEDDED_CONFIG=y;DIND_IMAGE=mirantis/kubeadm-dind-cluster:stable` in it, but perhaps it should only be updated on master when `stable` branch changes, so as to avoid confusion (newer dind script may not work with older image). So, I'm creating the `stable` branch then and adding this script?

**Comment by rpothier** [comment]
That would be great! What is the process if we needed to update the mirantis/kubeadm-dind-cluster:stable image?

**Comment by ivan4th** [comment]
I think I'll add option `-stable` to `build/genfixed.sh` that will fetch `dind-cluster.sh` script from stable branch, patch it to use `mirantis/kubeadm-dind-cluster:stable` image and write as `fixed/dind-cluster-stable.sh`. We can then update it on master every time `stable` branch is updated.

**Comment by ivan4th** [comment]
The best solution would be to get rid of `fixed/` altogether, packing almost all of kdc inside the image and leaving a very simple wrapper script to run kdc that would just start a container using dind image, like `docker run --rm --privileged --pid=host mirantis/kubeadm-dind-cluster:...` The container then would use nsenter to locate the docker socket (just mounting it via `-v` will not work on Mac) and then do all of `dind-cluster.sh` work. But right now I don't have enough time to implement that approach, maybe later... For now, I think what I wrote in the prev comment would do, what do you think?

**Comment by rpothier** [comment]
Yes, I think what was in the prev comment will do for now, thanks.

**Comment by leblancd** [comment]
@ivan4th Your previous comment seems like it might work for us, but one concern that I have is that we're using the gce-startup.sh script instead of using e.g. dind-cluster-stable.sh directly (our test run on GCE instances). Would we need a separate gce-startup-stable.sh script, or would there be a way to have the standard gce-startup.sh script call into dind-cluster-stable.sh?
Also, I like your second solution as a possible best approach! Instead of using nsenter to locate the docker socket, could the user just use a '-v /var/run/docker.sock:/var/run/docker.sock' argument in docker run?

**Comment by ivan4th** [comment]
Hi, I've added the new fixed script here: https://github.com/Mirantis/kubeadm-dind-cluster/blob/stable-fixed-version/fixed/dind-cluster-stable.sh
It uses `mirantis/virtlet:stable` image built from this branch: https://github.com/Mirantis/kubeadm-dind-cluster/tree/stable
Please confirm that this is what you need and I'll merge this new fixed script into master.

As of `-v /var/run/docker.sock:/var/run/docker.sock` from the last comment, it would be nice to be able to use it, but alas, it's not compatible with Mac Docker where `-v` has different meaning. In any case, getting rid of fixed scripts is perhaps one of the first things we should do shortly after upstreaming kdc.

**Comment by leblancd** [comment]
@ivan4th - Thanks! I'd like to test this with our CI tests (will try this Thursday, if that's okay), but from a brief look at the script, it appears that it will do what we need!

**Comment by leblancd** [comment]
@ivan4th - I haven't tested the new script yet in our CI environment. However, one thing to note is that our CI creates K8s clusters on GCE using the gce-setup.sh script. So to used the new 'dind-cluster-stable.sh' script in our setup, we would have to either:
(a) (Short term) Add a modified version of the gce-setup.sh script to our CI script that does everything gce-setup.sh does, except that at the very bottom, it calls the new 'dind-cluster-stable.sh' rather than 'dind-cluster.sh'.
(b) (Long term) Add a new 'gce-setup-stable.sh' script to the Mirantis repo that calls 'dind-cluster-stable.sh'???

**Comment by rpothier** [comment]
@leblancd @ivan4th For gce-setup, how about  we create a config file called config-stable.sh 
and put in in the fixed directory in the stable-fixed-version branch. 
In that file we set the env variables for the CI environment.
In the CI scripts, we set EMBEDDED_CONFIG=y, 
and run "cp fixed\config-stable.sh config.sh" before calling gce-setup.sh
We can then remove DIND_IMAGE and a few other env variables from the CI script also.

**Comment by rpothier** [comment]
For the above comment, I pushed up the changes for review.
https://github.com/Mirantis/kubeadm-dind-cluster/pull/123
